### PR TITLE
feat(Close #6): implement auth api

### LINKED ISSUE #6: implement auth api

### PR DESCRIPTION
**Pull request by khanghuynh92**

### CODE REVIEW

File: app.json [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    "year": "2017",
+    "language": "us",
+    "cookieId": "dc7161be3dbf2250c8954e560cc35061",
+    "author": {
+    	"name": "Gorillab",
+    	"email": "hello@gorillab.co",
+    	"url": "http://gorillab.co",
+    	"location": "Saigon"
+    }
+}
```
Comment (phatpham9) [inline app.json]: use `dotenv` to store environment variables instead https://github.com/motdotla/dotenv

File: config/auth.js [hunk 2]
```diff
@@ -0,0 +1,5 @@
+export const facebookAuth = {
+    clientID: '1922383334663210', // your App ID
+    clientSecret: '23d21a8e55c42af2fe765153490a8e16', // your App Secret
+    callbackURL: 'http://localhost:3000/api/v1/auth/facebook/callback'
+}
```
Comment (phatpham9) [inline config/auth.js]: store these env variables in `.env` file

File: config/database.js [hunk 3]
```diff
@@ -2,7 +2,7 @@ const host = process.env.DB_HOST || 'localhost';
 const port = process.env.DB_PORT || '27017';
 const database = process.env.DB_NAME || 'reader-api-dev';
```
Comment (phatpham9) [inline config/database.js]: remove all default variables

File: config/index.js [hunk 4]
```diff
@@ -1,2 +1,5 @@
+const envs = ['development', 'production', 'test']
+
 export const host = process.env.HOST || 'localhost';
 export const port = process.env.PORT || '3000';
```
Comment (phatpham9) [inline config/index.js]: remove all default variables

File: config/session.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+    cookie: {
+        path: '/',
+        httpOnly: true,
+        secure: false,
+        maxAge: 86400000    // 1 hour
+    },
+    resave: false,  //don't save session if unmodified
+    saveUninitialized: true, // don't create session until something stored
+    rolling: true   //Force a cookie to be set on every response. This resets the expiration date.
+};
```
Comment (phatpham9) [inline config/session.js]: remove this file. move this config to `index.js`

Comment (phatpham9) [reply config/session.js]: remove default variables also

File: controllers/AuthService.js [hunk 6]
```diff
@@ -1,56 +1,27 @@
...
-};
-  if (Object.keys(examples).length > 0) {
-    res.setHeader('Content-Type', 'application/json');
-    res.end(JSON.stringify(examples[Object.keys(examples)[0]] || {}, null, 2));
-  } else {
-    res.end();
-  }
-}
+import httpStatus from 'http-status';
+var passport = require('passport');
```
Comment (phatpham9) [inline controllers/AuthService.js]: should use `import`

File: middlewares/mongooseDefaultFields.js [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+      by: {
+        type: Schema.ObjectId,
+        ref: 'User',
+        select: false
+      }
+    }
+  };
+
+  if (field && typeof field === 'object' && !Array.isArray(field)) {
+    _.extend(fields, field);
```
Comment (phatpham9) [inline middlewares/mongooseDefaultFields.js]: avoid using lodash too much. use js `object.assign()` function instead

File: middlewares/mongooseDocExtend.js [hunk 8]
```diff
@@ -0,0 +1,17 @@
+var _ = require('lodash');
```
Comment (phatpham9) [inline middlewares/mongooseDocExtend.js]: use const

File: middlewares/mongooseDocExtend.js [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+module.exports = function mongooseDocExtend(schema) {
+  // extend updates
+  schema.method('extend', function(source) {
+    var self = this;
+
+    // extend doc
+    _.extend(self, source);
```
Comment (phatpham9) [inline middlewares/mongooseDocExtend.js]: use object.assign

File: middlewares/mongooseDocMethodsOverride.js [hunk 10]
```diff
@@ -0,0 +1,103 @@
+var _ = require('lodash');
```
Comment (phatpham9) [inline middlewares/mongooseDocMethodsOverride.js]: replace all var with const/let

File: middlewares/mongooseDocExtend.js [hunk 11]
```diff
@@ -0,0 +1,17 @@
+var _ = require('lodash');
+
+module.exports = function mongooseDocExtend(schema) {
```
Comment (phatpham9) [inline middlewares/mongooseDocExtend.js]: use es6 pls

Comment (khanghuynh92) [inline middlewares/mongooseDocExtend.js]: mongoose plugin can not use for es6

**Comment by khanghuynh92** [comment]
Updated!
@phatpham9

File: config/index.js [hunk 12]
```diff
@@ -1,5 +1,18 @@
...
+    cookie: {
+        path: '/',
+        httpOnly: true,
+        secure: false,
+        maxAge: 86400000    // 1 hour
+    },
+    resave: false,  //don't save session if unmodified
+    saveUninitialized: true, // don't create session until something stored
+    rolling: true   //Force a cookie to be set on every response. This resets the expiration date.
+}
```
Comment (phatpham9) [inline config/index.js]: I meant move all config to app index.js. we use dotenv to store env variables so no need to have this config file anymore

File: config/database.js [hunk 13]
```diff
@@ -1,10 +1,8 @@
-const host = process.env.DB_HOST || 'localhost';
-const port = process.env.DB_PORT || '27017';
-const database = process.env.DB_NAME || 'reader-api-dev';
+require('dotenv').config();
```
Comment (phatpham9) [inline config/database.js]: why require dotenv here?

Comment (phatpham9) [reply config/database.js]: just require it once in app index.js

**Comment by phatpham9** [comment]
naming scheme:
- class name: PascalCase such as `import Mongoose from 'mongoose'`, `import Passport from 'passport'`
- variable name: camelCase such as `const schema = Mongoose.schema`

**Comment by phatpham9** [comment]
naming scheme:
- class name: PascalCase such as `import Mongoose from 'mongoose'`, `import Passport from 'passport'`
- variable name: camelCase such as `const schema = Mongoose.schema`

File: config/database.js (hunk 13 above)
Comment (khanghuynh92) [reply config/database.js]: I don't know why the config folder don't understand env variables. although the app index.js file declared require('dotenv').config() before calling the config file.

**Comment by phatpham9** [comment]
@khanghuynh92 commit the empty `.env` file
and rename all classes, variables & functions following this naming scheme pls:
- class name: PascalCase such as import Mongoose from 'mongoose', import Passport from 'passport'
- variable/function name: camelCase such as const schema = Mongoose.schema

**Comment by khanghuynh92** [comment]
Updated! 
@phatpham9

File: .env [hunk 14]
```diff
@@ -0,0 +1,17 @@
...
+DB_PORT=27017
+DB_NAME=reader-api-dev
+DB_DEBUG=false
+
+SESSION_SECRET=readerapi
+
+
+FACEBOOK_CLIENT_ID=1922383334663210
+FACEBOOK_CLIENT_SECRET=23d21a8e55c42af2fe765153490a8e16
+FACEBOOK_CALLBACK=http://localhost:3000/api/v1/auth/facebook/callback
```
Comment (phatpham9) [inline .env]: Remove all values pls. I said empty .env file. This will cause credential data leaked.

Comment (khanghuynh92) [reply .env]: if remove all values in .env then travis scripts not satifying, right ?

Comment (phatpham9) [reply .env]: remove it. travis-ci has env variables settings. @tangkhanhphu will handle it.

Comment (khanghuynh92) [reply .env]: updated!

**Comment by khanghuynh92** [comment]
Updated!
@phatpham9

**Pull request merged**